crescent/internal/handle_resp.py

```diff
@@ -58,9 +59,35 @@ async def handle_resp(event: InteractionCreateEvent):
 
     command = _get_command(bot, name, interaction.guild_id, group, sub_group)
     ctx = Context._from_command_interaction(interaction)
-    callback_params = _options_to_kwargs(interaction, options)
+    callback_options = _options_to_kwargs(interaction, options)
 
-    await command.callback(ctx, **callback_params)
+    # ---------------
+    # Note on Copying
+    # ---------------
+    # To prevent the user accidentally mutating callback_options, shallow copies are
+    # passed to the user. Copies are only made when:
+    # 1. There is at least one interaction hook
+    # 2. The previous hook mutated the copy of callback_options
+
+    if command.interaction_hooks:
+        options_copy = copy(callback_options)
+
+    for hook in command.interaction_hooks:
+        if options_copy != callback_options:
+            options_copy = copy(callback_options)
+
+        hook_res = await hook(ctx, options_copy)
+
+        if hook_res:
+            if hook_res.options:
+                callback_options = hook_res.options
```

Comment:
What's the point of allowing them to return a new options dict if it just gets overriden? Or is the idea that only the last hook can actually mutate anything?

Reply:
It passes a copy so you can't accidentally mutate options. You have to return a new options dict for anything to change. I did a few tests and this should be comparable speed or faster than passing kwargs so I don't think copy will cause performance issues.